superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx

```diff
@@ -0,0 +1,230 @@
+   `}
+`;
+
+export type DrillDetailMenuItemsProps = {
+  chartId: number;
+  formData: QueryFormData;
+  filters?: BinaryQueryObjectFilterClause[];
+  isContextMenu?: boolean;
+  onSelection?: () => void;
+  onClick?: (event: MouseEvent) => void;
+};
+
+const DrillDetailMenuItems = ({
+  chartId,
+  formData,
+  filters = [],
+  isContextMenu = false,
+  onSelection = () => null,
+  onClick = () => null,
+  ...props
+}: DrillDetailMenuItemsProps) => {
+  const [modalFilters, setFilters] = useState<BinaryQueryObjectFilterClause[]>(
+    [],
+  );
+
+  const [showModal, setShowModal] = useState(false);
+  const openModal = useCallback(
+    (filters, event) => {
+      onClick(event);
+      onSelection();
+      setFilters(filters);
+      setShowModal(true);
+    },
+    [onClick, onSelection],
+  );
+
+  const closeModal = useCallback(() => {
+    setShowModal(false);
+  }, []);
+
+  // Check for Behavior.DRILL_TO_DETAIL to tell if plugin handles the `contextmenu`
+  // event for dimensions.  If it doesn't, tell the user that drill to detail by
+  // dimension is not supported.  If it does, and the `contextmenu` handler didn't
+  // pass any filters, tell the user that they didn't select a dimension.
+  const handlesDimensionContextMenu = useMemo(
+    () =>
+      getChartMetadataRegistry()
+        .get(formData.viz_type)
+        ?.behaviors.find(behavior => behavior === Behavior.DRILL_TO_DETAIL),
+    [formData.viz_type],
+  );
+
+  // Check chart plugin metadata to see if chart's current configuration lacks
+  // aggregations, in which case Drill to Detail should be disabled.
+  const noAggregations = useMemo(
+    () =>
+      getChartMetadataRegistry()
+        .get(formData.viz_type)
+        ?.noAggregations?.(formData) ?? false,
+    [formData],
+  );
+
+  const drillToDetail = noAggregations ? (
+    <DisabledMenuItem {...props} key="drill-detail-no-aggregations">
+      {t('Drill to detail')}
+      <DisabledMenuItemTooltip
+        title={t(
+          'Drill to detail is disabled because this chart does not group data by dimension value.',
+        )}
+      />
+    </DisabledMenuItem>
+  ) : (
+    <Menu.Item
+      {...props}
+      key="drill-detail-no-filters"
+      onClick={openModal.bind(null, [])}
+    >
+      {t('Drill to detail')}
+    </Menu.Item>
+  );
+
+  let drillToDetailBy;
+  if (noAggregations) {
+    drillToDetailBy = (
+      <DisabledMenuItem {...props} key="drill-detail-by-no-aggregations">
+        {t('Drill to detail by')}
+      </DisabledMenuItem>
+    );
+  } else {
+    drillToDetailBy = (
+      <DisabledMenuItem {...props} key="drill-detail-by-chart-not-supported">
+        {t('Drill to detail by')}
+        <DisabledMenuItemTooltip
+          title={t(
+            'Drill to detail by value is not yet supported for this chart type.',
+          )}
+        />
+      </DisabledMenuItem>
+    );
+
+    if (handlesDimensionContextMenu) {
+      if (filters?.length) {
```

Comment:
This is the third level if clause, should optimize readability.

Reply:
Moved all conditionals to the top level – does that look better?